### DISCUSSION TITLE
ScriptReference/Physics.CapsuleCast

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#316] [Notes]
# ScriptReference/Physics.CapsuleCast

https://docs.unity3d.com/ScriptReference/Physics.CapsuleCast.html

**Comment by vertxxyz** [comment]
`@cjacobwade` previously noted:
> The example code shown here is misleading. The parameter info above notes Point1 and Point2 must be at the center of the spheres at the ends of the capsule, while the example code uses CharacterController.height which is the length of a capsule including Radius of the spheres on each end.

And provided tweaked code that solved that issue and attempted to take scaling into account.
I've found that there's a number of extra things to take into account in regards to CharacterControllers. They cannot be rotated, and scaling is handled in a non-standard manner. Here is my code that robustly handles those cases:
```csharp
Transform transform = characterController.transform;
Vector3 scale = transform.lossyScale;
float radiusScale = Mathf.Max(Mathf.Abs(scale.x), Mathf.Abs(scale.z));
Radius = characterController.radius * radiusScale;
float offset = Mathf.Max(characterController.height * 0.5f * Mathf.Abs(scale.y), Radius) - Radius;
Vector3 center = transform.TransformPoint(characterController.center);
SpherePosition1 = new Vector3(center.x, center.y - offset, center.z);
SpherePosition2 = new Vector3(center.x, center.y + offset, center.z);
// Output: Radius, SpherePosition1, SpherePosition2
```

**Reply by vertxxyz** [reply]
Similar logic, but for CapsuleCollider:
```csharp
Transform transform = collider.transform;
Vector3 scale = transform.lossyScale;
float radiusScale = Mathf.Max(Mathf.Abs(scale.x), Mathf.Abs(scale.z));
Radius = collider.radius * radiusScale;
Vector3 center = transform.TransformPoint(collider.center);
float offsetY = Mathf.Max(collider.height * 0.5f * Mathf.Abs(scale.y), Radius) - Radius;
Vector3 offset = transform.TransformDirection(new Vector3(0, offsetY, 0));
SpherePosition1 = center - offset;
SpherePosition2 = center + offset;
// Output: Radius, SpherePosition1, SpherePosition2
```
